core/sdk-core/src/main/java/software/amazon/awssdk/core/endpointdiscovery/EndpointDiscoveryRefreshCache.java

```diff
@@ -87,6 +90,64 @@ public URI get(String accessKey, EndpointDiscoveryRequest request) {
         return endpoint.endpoint();
     }
 
+    public CompletableFuture<URI> getAsync(String accessKey, EndpointDiscoveryRequest request) {
+        key = accessKey;
+
+        // Support null (anonymous credentials) by mapping to empty-string. The backing cache does not support null.
+        if (key == null) {
+            key = "";
+        }
+
+        if (request.cacheKey().isPresent()) {
+            key = key + ":" + request.cacheKey().get();
+        }
+
+        EndpointDiscoveryEndpoint endpoint = cache.get(key);
+
+        if (endpoint == null) {
+            if (request.required()) {
+                return discoverEndpoint(request).handle(
+                    (endpointDiscoveryEndpoint, throwable) -> {
+                        if (throwable != null) {
+                            if (throwable instanceof InterruptedException) {
+                                Thread.currentThread().interrupt();
+                                throw EndpointDiscoveryFailedException.create(throwable);
+                            }
+                            if (throwable instanceof ExecutionException
+                                    || throwable instanceof CompletionException) {
+                                throw EndpointDiscoveryFailedException.create(throwable.getCause());
+                            }
+                            throw new RuntimeException("new exception");
+                        }
+                        return cache.computeIfAbsent(
+                            key, k -> endpointDiscoveryEndpoint
+                        ).endpoint();
+                    });
+            }
+            EndpointDiscoveryEndpoint tempEndpoint = EndpointDiscoveryEndpoint.builder()
+                                                                              .endpoint(request.defaultEndpoint())
+                                                                              .expirationTime(Instant.now().plusSeconds(60))
+                                                                              .build();
+
+            EndpointDiscoveryEndpoint previousValue = cache.putIfAbsent(key, tempEndpoint);
+            if (previousValue != null) {
+                // Someone else primed the cache. Use that endpoint (which may be temporary).
+                return CompletableFuture.completedFuture(previousValue.endpoint());
+            } else {
+                // We primed the cache with the temporary endpoint. Kick off discovery in the background.
+                refreshCacheAsync(request, key);
+            }
+            return CompletableFuture.completedFuture(tempEndpoint.endpoint());
+        }
+
+        if (endpoint.expirationTime().isBefore(Instant.now())) {
+            cache.put(key, endpoint.toBuilder().expirationTime(Instant.now().plusSeconds(60)).build());
+            refreshCacheAsync(request, key);
+        }
+
+        return CompletableFuture.completedFuture(endpoint.endpoint());
+    }
```

Comment:
Same here, can we share the code used in sync?

Reply:
done. I have modified this block